AmebaZ2: Add soc src code

jerry-tang-realsil opened the pull request:
Add soc src code for rtl8720c

Signed-off-by: Jerry_tang <jerry_tang@realsil.com.cn>

## Summary
add soc-rtl8720c src code
## Impact
no
## Testing
ok

acassis commented:
This code derive from GPL code, see PR #4066. *Please* get the approval (SGA) from Realtek, otherwise we cannot include it on NuttX.

jerry-tang-realsil commented:
I’m sorry but I don’t understand what the approval should be.
Could you please show me what the approval is? Is there any example?

发件人: Alan Carvalho de Assis ***@***.***
发送时间: 2021年7月6日 19:33
收件人: apache/incubator-nuttx ***@***.***>
抄送: 唐帅帅 ***@***.***>; Author ***@***.***>
主题: Re: [apache/incubator-nuttx] AmebaZ2: Add soc src code (#4076)


This code derive from GPL code, see PR #4066<https://github.com/apache/incubator-nuttx/pull/4066>. Please get the approval (SGA) from Realtek, otherwise we cannot include it on NuttX.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/apache/incubator-nuttx/pull/4076#issuecomment-874683612>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AP2FFPESUJSIRWN4FXVMUTTTWLSXJANCNFSM473V3NFA>.

------Please consider the environment before printing this e-mail.

davids5 commented:
@jerry-tang-realsil 

Please see https://www.apache.org/licenses/contributor-agreements.html

once Realtek submits a CCLA: Corporate Contributor License Agreement then the code can come in.

jerry-tang-realsil commented:
I added a wrong license before. 
Below license is from our release code.
So does it still need to sign a SGA? 

/ ******************************************************************************
 *
 * Copyright(c) 2007 - 2017 Realtek Corporation. All rights reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the License); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an AS IS BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ******************************************************************************/

acassis commented on arch/arm/src/rtl8720c/amebaZ.c:
```diff
@@ -0,0 +1,47 @@
+/****************************************************************************
+ * arch/arm/src/rtl8720c/amebaZ.c
```
Don't use lower case mixed with upper case file names. Please rename the file to amebaz.c

jerry-tang-realsil replied:
ok，i'll modify it

acassis commented on arch/arm/src/rtl8720c/amebaZ.c:
```diff
@@ -0,0 +1,47 @@
+
+void ameba_reset(int status)
+{
+  hal_sys_set_fast_boot(0, 0);
+  hal_misc_rst_by_wdt();
```
Is it possible to avoid using the HAL? It is better to have a slim implementation here than using a blackbox HAL.

acassis commented:
> I added a wrong license before.
> Below license is from our release code.
> So does it still need to sign a SGA?
> 
> / ******************************************************************************
> *
> 
>     * Copyright(c) 2007 - 2017 Realtek Corporation. All rights reserved.
> 
>     * SPDX-License-Identifier: Apache-2.0
> 
>     * Licensed under the Apache License, Version 2.0 (the License); you may
> 
>     * not use this file except in compliance with the License.
> 
>     * You may obtain a copy of the License at
> 
>     * http://www.apache.org/licenses/LICENSE-2.0
> 
>     * Unless required by applicable law or agreed to in writing, software
> 
>     * distributed under the License is distributed on an AS IS BASIS, WITHOUT
> 
>     * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> 
>     * See the License for the specific language governing permissions and
> 
>     * limitations under the License.
> 
> 
> ******************************************************************************/

Yes, it is a protection for this project. It will avoid someone claiming in the future that we are stealing code from Realtek. If you work for Realtek it should be fast to get this approach, we get it here at Espressif in less than 3 days. Please ask some Realtek director to submit the documents.

acassis commented:
Please keep is as Draft to avoid someone accidentally merging it.

jerry-tang-realsil commented:
First, Realtek has ever signed Apache License, so we still need to submit relative documents?
Then, these code's license was added by mistake. It need not add realtek licnese actually.
Above all, I don't think there is need to discuss the problem of license.

acassis commented:
> First, Realtek has ever signed Apache License, so we still need to submit relative documents?
> Then, these code's license was added by mistake. It need not add realtek licnese actually.
> Above all, I don't think there is need to discuss the problem of license.

@justinmclean what do you think in this case? jerry-tang-realsil submitted a code as GPL, then he said it was a mistake, now he says Realtek already has signed a SGA with Apache (I suppose for other project) and he thinks we don't need to discuss the problem of the license. The issue is not the license "per se", but the fact he brought existent code that seams to come from other project and/or company (including code from Xiaomi that he removed the Copyright). I don't see someone in the future saying we are stealing code from other project and adding on NuttX. So, do you think we can ignore these facts and just let it in?

jerry-tang-realsil commented:
Yeah, some code is from xiaomi. But xiaomi has told us that license just refer to others, such as esp32. And not all of the code added xiaomi's license. So I modified it but made a mistake that  I added an extra wrong license. It's because I misunderstand the requirement of license.


发自我的iPhone

在 2021年7月9日，下午8:07，Alan Carvalho de Assis ***@***.***> 写道：

﻿

First, Realtek has ever signed Apache License, so we still need to submit relative documents?
Then, these code's license was added by mistake. It need not add realtek licnese actually.
Above all, I don't think there is need to discuss the problem of license.

@justinmclean<https://github.com/justinmclean> what do you think in this case? jerry-tang-realsil submitted a code as GPL, then he said it was a mistake, now he says Realtek already has signed a SGA with Apache (I suppose for other project) and he thinks we don't need to discuss the problem of the license. The issue is not the license "per se", but the fact he brought existent code that seams to come from other project and/or company (including code from Xiaomi that he removed the Copyright). I don't see someone in the future saying we are stealing code from other project and adding on NuttX. So, do you think we can ignore these facts and just let it in?

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub<https://github.com/apache/incubator-nuttx/pull/4076#issuecomment-877138682>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AP2FFPASXGGZBGFXEGKJHDTTW3Q7LANCNFSM473V3NFA>.

------Please consider the environment before printing this e-mail.

acassis commented:
yes @jerry-tang-realsil I understand your point, but we cannot take the risk. Because we don't know for sure from where this code came, the right thing to do now is get approval from Realtek since the original GPL code had their Copyright. I hope you understand it.

jerry-tang-realsil commented:
Do you mean that I should add realtek apache license?

发自我的iPhone

在 2021年7月9日，下午9:52，Alan Carvalho de Assis ***@***.***> 写道：

﻿

yes @jerry-tang-realsil<https://github.com/jerry-tang-realsil> I understand your point, but we cannot take the risk. Because we don't know for sure from where this code came, the right thing to do now is get approval from Realtek since the original GPL code had their Copyright. I hope you understand it.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/apache/incubator-nuttx/pull/4076#issuecomment-877205342>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AP2FFPEPTVB45POTFGDZWNDTW35IDANCNFSM473V3NFA>.

------Please consider the environment before printing this e-mail.

justinmclean commented:
A SGA only covers code that has been donated in that SGA it doesn't cover all code from a company. So first question is has this exact piece of code being donated by a software grant? If that is not the case what license is the code under? If the code is under GPL then we can't accept it as GPL is not compatible with the Apace license.

justinmclean commented:
I'll also add that any code where the header has been altered or the copyright removed without the permission of the owner should be rejected.

acassis commented:
@justinmclean Mr. @jerry-tang-realsil works for Realtek, but the original code was submitted here as GPL. I think Realtek as owner of the original code could donate that code to be released under Apache License. Do you think the CCLA is enough in this case or should we ask Realtek the SGA to grant permission of this code?

jerry-tang-realsil commented:
I may explain again:


These code is based on realtek's code. Xiaomi transplanted it with realtek's help. It's still realtek's duty to maintain in future.

So, xiaomi require realtek to donate these code to nuttx, and realtek agreed.


Realtek 's code is not relative to GPL.


I also think it's ok to add Apache License. Xiaomi also advised us to modify the license, refer to others that have donated to nuttx.


________________________________
发件人: Alan Carvalho de Assis ***@***.***>
发送时间: 2021年7月9日 22:51:44
收件人: apache/incubator-nuttx
抄送: 唐帅帅; Mention
主题: Re: [apache/incubator-nuttx] AmebaZ2: Add soc src code (#4076)


@justinmclean<https://github.com/justinmclean> Mr. @jerry-tang-realsil<https://github.com/jerry-tang-realsil> works for Realtek, but the original code was submitted here as GPL. I think Realtek as owner of the original code could donate that code to be released under Apache License. Do you think the CCLA is enough in this case or should we ask Realtek the SGA to grant permission of this code?

―
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/apache/incubator-nuttx/pull/4076#issuecomment-877245362>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AP2FFPHQYCY6VRZZYLPB7SLTW4EIBANCNFSM473V3NFA>.

------Please consider the environment before printing this e-mail.

acassis commented:
> I may explain again: These code is based on realtek's code. Xiaomi transplanted it with realtek's help. It's still realtek's duty to maintain in future. So, xiaomi require realtek to donate these code to nuttx, and realtek agreed. Realtek 's code is not relative to GPL. I also think it's ok to add Apache License. Xiaomi also advised us to modify the license, refer to others that have donated to nuttx.
> […](#)
> ________________________________ 发件人: Alan Carvalho de Assis ***@***.***> 发送时间: 2021年7月9日 22:51:44 收件人: apache/incubator-nuttx 抄送: 唐帅帅; Mention 主题: Re: [apache/incubator-nuttx] AmebaZ2: Add soc src code (#4076) @justinmclean<https://github.com/justinmclean> Mr. @jerry-tang-realsil<https://github.com/jerry-tang-realsil> works for Realtek, but the original code was submitted here as GPL. I think Realtek as owner of the original code could donate that code to be released under Apache License. Do you think the CCLA is enough in this case or should we ask Realtek the SGA to grant permission of this code? ― You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub<[#4076 (comment)](https://github.com/apache/incubator-nuttx/pull/4076#issuecomment-877245362)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AP2FFPHQYCY6VRZZYLPB7SLTW4EIBANCNFSM473V3NFA>.
> ------Please consider the environment before printing this e-mail.

Thank you @jerry-tang-realsil for this explanation. I think we are reaching a better understanding of this situation. Is the original code from Realtek available publicly under Apache or other compatible license? Is so, could you give us the link to it? Case it is just internal code inside Realtek company, then I think the SGA is needed, because we cannot prove that original code license was compatible with Apache license.

patacongo commented:
> Is the original code from Realtek available publicly under Apache or other compatible license? Is so, could you give us the link to it? Case it is just internal code inside Realtek company, then I think the SGA is needed, because we cannot prove that original code license was compatible with Apache license.

If Realtek owns of code and legitimately holds the copyright, then they may change the license as they see fit.  Right?    If that condition is true that I cannot imagine what the issue with licensing could possibly be.

With regard to the SGA, my understanding is that this depends on a subjective understanding of the scope of the change.  Is that right?  This is beyond my ability to suggest anything credible.  SGAs are problematic for some businesses (or at the very least, time consuming).  I suppose that you could argue either way that architecture support is broad or narrow scope.  I would end to go with narrow scope since it does not reflect any core technology that is critical to the integrity of the OS; it simply extends existing support to the Realtek products.

If the Realtek support will continue to develop over time, like ESP32 and Spresense, then we would need an SGA.

acassis commented:
> > Is the original code from Realtek available publicly under Apache or other compatible license? Is so, could you give us the link to it? Case it is just internal code inside Realtek company, then I think the SGA is needed, because we cannot prove that original code license was compatible with Apache license.
> 
> If Realtek owns of code and legitimately holds the copyright, then they may change the license as they see fit. Right? If that condition is true that I cannot imagine what the issue with licensing could possibly be.
> 
> With regard to the SGA, my understanding is that this depends on a subjective understanding of the scope of the change. Is that right? This is beyond my ability to suggest anything credible. SGAs are problematic for some businesses (or at the very least, time consuming). I suppose that you could argue either way that architecture support is broad or narrow scope. I would end to go with narrow scope since it does not reflect any core technology that is critical to the integrity of the OS; it simply extends existing support to the Realtek products.
> 
> If the Realtek support will continue to develop over time, like ESP32 and Spresense, then we would need an SGA.

Ok Greg, makes sense! If in the future Realtek assign the SGA it will be ok, otherwise if someone complaint about the origin of this code we could remove it, although I think this is not the best option.

patacongo commented:
Whether an SGA is signed or not, Realtek needs to be cognizant of the fact that they are granting the code the Apache Software Foundation.  The ASF now owns the code; the ASF now holds the copyright.  Other uses of this code by Realtek in the future could not violate the ASF's copyright.  I hope that we do not encounter such a legal dilemma.

patacongo commented:
> 
> 
> Whether an SGA is signed or not, Realtek needs to be cognizant of the fact that they are granting the code the Apache Software Foundation. The ASF now owns the code; the ASF now holds the copyright. Other uses of this code by Realtek in the future could not violate the ASF's copyright. I hope that we do not encounter such a legal dilemma. Signing the SGA would make this gift to the ASF much more explicit.

I might have that wrong.  I sometimes think I know more than I really do.  The actual test of the file header is:

     * Licensed to the Apache Software Foundation (ASF) under one or more
     * contributor license agreements.  See the NOTICE file distributed with
     * this work for additional information regarding copyright ownership.  The
     * ASF licenses this file to you under the Apache License, Version 2.0 (the
     * "License"); you may not use this file except in compliance with the
     * License.  You may obtain a copy of the License at ....

This is probably why it is important to sign the contributor agreement (SGA and ICLA).  Nothing was added to the NOTICE file, right?  The alternative that Realtek probably really wanted was to release this as third party code.

- https://www.apache.org/legal/resolved.html 
- https://www.apache.org/foundation/license-faq.html#Apply-My-Software

acassis commented:
Yes, that was my worrying. It is difficult to track back this code's origin and nobody from Realtek submitted an ICLA, CCLA or SGA to the Apache related to the NuttX. But Ok, let believe they will submit it soon, maybe Xiaomi could ask them to do so.

xiaoxiang781216 commented:
Sure, I will talk with Realtek.

jerry-tang-realsil commented:
@ Xiang Xiao

All right, we would like to discuss the issue and modify according to your requirement.

Sorry for the mistake.

Please email us or WeChat the date if need a meeting.

________________________________
发件人: Xiang Xiao ***@***.***>
发送时间: 2021年7月10日 16:38:23
收件人: apache/incubator-nuttx
抄送: 唐帅帅; Mention
主题: Re: [apache/incubator-nuttx] AmebaZ2: Add soc src code (#4076)


Sure, I will talk with Realtek.

―
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/apache/incubator-nuttx/pull/4076#issuecomment-877597629>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AP2FFPCAN34C46IIZY72PZTTXABH7ANCNFSM473V3NFA>.

------Please consider the environment before printing this e-mail.
